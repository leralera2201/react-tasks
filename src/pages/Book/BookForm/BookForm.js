import React, { Component } from "react";
import { Button, Form } from "reactstrap";
import { connect } from "react-redux";

import Field from "@components/Field";

import { createBookStart, updateBookStart } from "../actions/book.actions";

const fields = {
  title: {
    label: "Title",
    type: "text",
    value: "",
  },
  description: {
    label: "Description",
    type: "text",
    value: "",
  },
  excerpt: {
    label: "Excerpt",
    type: "text",
    value: "",
  },
  publishDate: {
    label: "Publish Date",
    type: "date",
    value: "",
  },
  pageCount: {
    label: "Page Count",
    type: "number",
    value: "",
  },
};

export class BookForm extends Component {
  state = {
    ...fields,
  };

  componentDidMount() {
    const {
      location: { state },
    } = this.props;
    if (state && state.book) {
      Object.entries(state.book).forEach(([fieldName, value]) => {
        this.setState({
          [fieldName]: {
            ...this.state[fieldName],
            value,
          },
        });
      });
    }
  }

  changeHandler = (event) => {
    const { name, value } = event.target;
    const data = this.state;
    data[name] = { ...data[name], value };

    this.setState({ ...data });
  };

  submitHandler = () => {
    const {
      createBook,
      updateBook,
      history,
      match: {
        params: { id },
      },
    } = this.props;
    const data = Object.entries(this.state).reduce((result, field) => {
      const [fieldName, { value }] = field;
      result[fieldName] = value;
      return result;
    }, {});
    if (id) {
      updateBook(data);
    } else {
      createBook(data);
    }
    this.setState({ ...fields });
    history.push({
      pathname: "/books",
      state: { withoutFetch: true },
    });
  };

  render() {
    const data = this.state;
    const {
      match: {
        params: { id },
      },
    } = this.props;
    return (
      <div className="container pt-5">
        <Form>
          {Object.entries(data).map(([fieldName, field]) => (
            <Field
              key={fieldName}
              type={field.type}
              label={field.label}
              name={fieldName}
              value={field.value}
              onChange={this.changeHandler}
            />
          ))}
          <Button onClick={this.submitHandler}>{id ? "Update" : "Add"}</Button>
        </Form>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  createBook: (book) => dispatch(createBookStart(book)),
  updateBook: (book) => dispatch(updateBookStart(book)),
});

export default connect(null, mapDispatchToProps)(BookForm);
