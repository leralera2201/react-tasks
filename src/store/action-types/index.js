export const ACTION_STATUS = {
  NOT_STARTED: "NOT_STARTED",
  IN_PROGRESS: "IN_PROGRESS",
  ERROR: "ERROR",
  SUCCESS: "SUCCESS",
};

export const createActionTypes = (actionTypeName) => ({
  START: `${actionTypeName}_START`,
  IN_PROGRESS: `${actionTypeName}_IN_PROGRESS`,
  SUCCESS: `${actionTypeName}_SUCCESS`,
  ERROR: `${actionTypeName}_ERROR`,
  RESET: `${actionTypeName}_RESET`,
});
