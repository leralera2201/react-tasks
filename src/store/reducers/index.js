import { combineReducers } from "redux";

import BooksReducer from "../../pages/Books/reducers/books.reducer";
import BookReducer from "../../pages/Book/reducers/book.reducer";

const rootReducer = combineReducers({
  books: BooksReducer,
  book: BookReducer,
});

export default rootReducer;
