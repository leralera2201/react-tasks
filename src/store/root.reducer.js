import { combineReducers } from "redux";
import BookReducer from "./reducers/book.reducer";

const rootReducer = combineReducers({
  books: BookReducer,
});

export default rootReducer;
