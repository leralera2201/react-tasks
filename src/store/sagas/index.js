import { all, fork } from 'redux-saga/effects';

import bookSagas from '../../pages/Book/sagas/book.sagas';

const combinedSagas = {
  ...bookSagas,
};

export default function* rootSaga() {
  yield all(Object.values(combinedSagas).map((saga) => fork(saga)));
}
