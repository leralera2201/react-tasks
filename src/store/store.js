import { createStore, applyMiddleware, compose } from "redux";
import createSagaMiddleware from "redux-saga";

import rootReducer from "./reducers";
import allSagas from "./sagas";

const sagaMiddleware = createSagaMiddleware();
const composeEnhancers =
  (typeof window !== "undefined" &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) ||
  compose;

const store = createStore(
  rootReducer,
  composeEnhancers(applyMiddleware(sagaMiddleware))
);

sagaMiddleware.run(allSagas);

export default store;
