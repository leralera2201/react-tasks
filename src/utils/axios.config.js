import axios from "axios";

const instance = axios.create({
  baseURL: "https://fakerestapi.azurewebsites.net",
});

export default instance;
