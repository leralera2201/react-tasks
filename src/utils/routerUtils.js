import React from 'react';
import { MemoryRouter } from 'react-router';

import App from "../App";

export const componentWithRouter = (route='/', component = <App/>) => <MemoryRouter initialEntries={[route]} initialIndex={0}>
    {component}
</MemoryRouter>