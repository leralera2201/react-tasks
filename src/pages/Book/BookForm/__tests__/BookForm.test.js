import React from 'react';
import {mount} from "enzyme";
import {Provider} from "react-redux";
import store from "../../../../store/store";
import {componentWithRouter} from "../../../../utils/routerUtils";
import {BookForm} from "../BookForm";
import {Button} from "reactstrap";


describe("BookForm", () => {

    const props = {
        match: {
            params: {}
        },
        location: {
            state: {}
        },
        createBook: jest.fn(),
        updateBook: jest.fn(),
        history: { push: jest.fn(), location: {}, listen: jest.fn() }
    }

    it('should have 5 inputs', () => {
        const wrapper = mount(<BookForm {...props} />);
        expect(wrapper.find('input').length).toEqual(5)
    });

    it('should change the state after change the input value', () => {
        const newValue = 'testing component';
        const wrapper = mount(<BookForm {...props} />)
        const input = wrapper.find('input').first();
        input.simulate('change', { target: { name: 'title', value: newValue }});

        expect(wrapper.state().title.value).toEqual(newValue);
    });

    it('should change state', () => {
        props.location.state = {
            book: {
                title: "d",
                description: "d",
                excerpt: "d",
                publishDate: "2000-20-20",
                pageCount: 2,
                id: '1'
            }
        }
        const wrapper = mount(<BookForm {...props} />);
        expect(wrapper.state().title.value).toEqual(props.location.state.book.title);
    });

    it('should change path', () => {
        const wrapper = mount(<BookForm {...props} />)
        const button = wrapper.find('button').first();
        button.simulate('click');
        expect(props.history.push.mock.calls[0][0].pathname).toEqual('/books');
    });

    it("should update book", () => {
        props.match.params.id = '1';
        const wrapper = mount(<BookForm {...props} />)
        const button = wrapper.find('button').first();
        button.simulate('click');
        expect(props.updateBook).toBeCalled();
    })

})