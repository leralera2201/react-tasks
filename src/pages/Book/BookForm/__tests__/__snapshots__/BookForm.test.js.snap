// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookForm renders correctly 1`] = `
<div
  className='container pt-5'
>
  <Form
    tag='form'
  >
    <Field
      key='title'
      label='Title'
      name='title'
      onChange={[Function]}
      type='text'
      value=''
    />
    <Field
      key='description'
      label='Description'
      name='description'
      onChange={[Function]}
      type='text'
      value=''
    />
    <Field
      key='excerpt'
      label='Excerpt'
      name='excerpt'
      onChange={[Function]}
      type='text'
      value=''
    />
    <Field
      key='publishDate'
      label='Publish Date'
      name='publishDate'
      onChange={[Function]}
      type='date'
      value=''
    />
    <Field
      key='pageCount'
      label='Page Count'
      name='pageCount'
      onChange={[Function]}
      type='number'
      value=''
    />
    <Button
      color='secondary'
      onClick={[Function]}
      tag='button'
    >
      Add
    </Button>
  </Form>
</div>
`;
